src/Discovery/DiscoveryDetails/DataDownloadList/ActionButtons/DownloadUtils/DownloadVariableMetadata/DownloadVariableMetadata.tsx

```diff
@@ -0,0 +1,104 @@
+import DownloadStatus from '../../../Interfaces/DownloadStatus';
+import { DiscoveryResource } from '../../../../../Discovery';
+
+interface IdataDictionaries {
+  value: {
+    [key: string]: any;
+  };
+}
+
+const DownloadVariableMetadata = async (
+  resourceInfo: DiscoveryResource,
+  setDownloadStatus: Function,
+) => {
+  const zip = new JSZip();
+  const studyID = resourceInfo._hdp_uid;
+  const projectTitle = resourceInfo.project_title;
+
+  const DOWNLOAD_FAIL_INFO: DownloadStatus = {
+    inProgress: false,
+    message: {
+      title: 'Download failed',
+      content: (
+        <React.Fragment>
+          <p>
+            Study with name:<strong> {projectTitle}</strong> cannot download
+            variable-level metadata with ID:
+            <strong> {studyID}</strong>.
+          </p>
+          <p>Please try again later and contact support.</p>
+        </React.Fragment>
+      ),
+      active: true,
+    },
+  };
+
+  const createUniqueDownloadErrorMsg = (key: string) => ({
+    ...DOWNLOAD_FAIL_INFO,
+    message: {
+      ...DOWNLOAD_FAIL_INFO.message,
+      content: (
+        <React.Fragment>
+          <p>
+            Data Dictionary with name <strong>{key}</strong> cannot download data
+            dictionary with name <strong>{key}</strong>.
+          </p>
+          <p>Please try again later and contact support.</p>
+        </React.Fragment>
+      ),
+    },
+  });
+
+  const fetchData = async (key: string, value: string) => new Promise((resolve, reject) => {
+    fetchWithCreds({ path: `${mdsURL}/${value}` }).then((statusResponse) => {
+      const { data } = statusResponse;
+      if (statusResponse.status !== 200 || !data) {
+        setDownloadStatus(createUniqueDownloadErrorMsg(key));
+        reject(new Error(`Issue with ${key}: ${value}`));
+      } else {
+        zip.file(key, JSON.stringify(data));
+        resolve(`Data resolved for ${key}: ${value}`);
+      }
+    });
+  });
+
+  const fetchDataForAllFiles = async (dataDictionaries: IdataDictionaries) => {
+    try {
+      await Promise.all(
+        Object.entries(dataDictionaries).map(([key, value]) => fetchData(key, value),
+        ),
+      ).then(() => {
+        zip.generateAsync({ type: 'blob' }).then((content) => {
+          FileSaver.saveAs(content, 'variable-metadata.zip');
+        });
+      });
+    } catch (error) {
+      console.error('Error fetching data:', error);
+    }
+  };
+
+  fetchWithCreds({ path: `${mdsURL}/${studyID}` }).then((statusResponse) => {
+    const { data } = statusResponse;
+    if (statusResponse.status !== 200 || !data) {
+      setDownloadStatus(DOWNLOAD_FAIL_INFO);
+    } else {
+      try {
+        const dataDictionaries: IdataDictionaries = data.data_dictionaries;
+        if (Object.keys(dataDictionaries).length !== 0) {
+          fetchDataForAllFiles(dataDictionaries);
+        }
+      } catch (error) {
+        setDownloadStatus(DOWNLOAD_FAIL_INFO);
```

Comment:
So you have set the download status by using `createUniqueDownloadErrorMsg()` when there is an error downloading a particular DD in `fetchData()`. And then you called `fetchData()` in `fetchDataForAllFiles` and surronded it with a try-catch block. Would that make this catch block here redundent?

Reply:
It is possible that _fetchDataForAllFiles_ will not execute if the _Object.keys(dataDictionaries).length !== 0_ is false (i.e. there's no data dictionaries associated with the study). In that case the catch block at line 95 executes and shows the warning. 

An example: Study  HDP00258 is not configured for Data Dictionaries, so its data.datadictionaries is undefined, causing the catch block to execute.
![output](https://github.com/uc-cdis/data-portal/assets/113449836/d130b6f0-1950-4817-833e-95c6feeb21b2)

An alternative approach would be to disable the button for these studies, but I think that would require an API call each time. Let me know what you think.

Reply:
I see, in this case, I think we can move this line that is getting DD keys/GUIDs from the input metadata https://github.com/uc-cdis/data-portal/pull/1476/files#diff-c6cbd176398b0ad195cfb42288fcf60a404070635158552da44e2e4442f40a89R91 into src/Discovery/DiscoveryDetails/DataDownloadList/ActionButtons/ActionButtons.tsx
So it will be something like

> const dataDictionaries: IdataDictionaries = data[discoveryConfig.features.exportToWorkspace.variableMetadataFieldName];
> const noVLMDData = !dataDictionaries || dataDictionaries.length == 0
> // use noVLMDData to disable/enable the download VLMD button at L82

In this way, we can disable the download VLMD button for studies that doesn't have VLMD attached to them in metadata without making additional API calls, and we can also simplify the try-catch blocks in this file

Reply:
This has been updated: the try catch block in _DownloadVariableMetadata_ has been removed. The logic for downloading the data dictionary info has been placed into a separate file _DownloadDataDictionaryInfo_ and is called in the _ActionsButtons_ component. This updates a state _dataDictionaryInfo_ with values for _noVariableMetadata_ and _dataDictionaries_ which is passed down as a parameter to the _DownloadVariableMetadata_ function. A unit test has been written for the new file and the unit tests for the other files have been updated as needed.